File: CTCI/Interview Questions/Python/2.6_palindrome.py
# Implement a function to check if a linked list is a palindrome.

# Page 64

import copy # For creating an object copy

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def fill(self, elements):
        if self.head:
            self.head = None # Doing this just to make the main function cleaner
        for element in elements:
            self.append(element)

    def append(self, data):
        if self.head is None:
            self.head = Node(data)
            current = self.head
            while current.next:
                current = current.next
            current.next = Node(data)

    def set_head(self, new_head):
        self.head = new_head

    def __str__(self):
        output_list = []
        current = self.head
        while current:
            output_list.append(current.data)
            current = current.next
        return ' '.join([str(ele) for ele in output_list])

# Solution 1
# Using a stack store store the first half of the palindrome, then popping each and checking with latter half of linked list.
# Solution assumes that size of linked list is not known.
# Time Complexity: O(n), Space Complexity: O(n)
def is_palindrome_1(ll):
    # If string is less than 1 character long
    if ll.head is None or ll.head.next is None:
        return True
    # If string is two characters long
    if ll.head.data == ll.head.next.data and ll.head.next.next is None:
        return True
    slow = ll.head
    fast = ll.head
    stack = []
    while True:
        stack.append(slow.data)
        slow = slow.next
        fast = fast.next.next
        # If character has an odd number of characters
        if fast.next is None:
            slow = slow.next
            break
        # If character has an even number of characters
        elif fast.next is not None and fast.next.next is None:
            stack.append(slow.data)
            slow = slow.next
            break
    while slow:
        if slow.data != stack.pop():
            return False
        slow = slow.next
    return True

def reverse_and_clone_ll(ll):
    rev_ll = copy.deepcopy(ll) # Cloning linked list. Deep copy means changes to the copied-from linked list won't have any effect on the copied linked list.
    current = rev_ll.head
    prev = None
    while current:
        next = current.next
        current.next = prev
        prev = current
        current = next
    rev_ll.set_head(prev)
    return rev_ll

def compare_linked_lists(ll, rev_ll):
    first = ll.head
    second = rev_ll.head
    fast = ll.head
    # Since we're looking for palindromes, traversing through the entire list isn't necessary. We stop when we're at the halfway point.
    while first and second:
        if first.data != second.data:
            return False
        first = first.next
        second = second.next
        if fast.next and fast.next.next:
            fast = fast.next.next
        else:
            break
    return True

# Solution 2
# Reversing linked list and check first half of each list
# Time Complexity: O(n), Space Complexity: O(n)
def is_palindrome_2(ll):
    rev_ll = reverse_and_clone_ll(ll)
    return compare_linked_lists(ll, rev_ll)


ll = LinkedList()
ll.fill([ch for ch in 'red rum sir is murder'.replace(' ', '')])
print(ll, '->', is_palindrome_1(ll))
print(ll, '->', is_palindrome_2(ll))

ll.fill([ch for ch in 'hello world lol'.replace(' ', '')])
print(ll, '->', is_palindrome_1(ll))
print(ll, '->', is_palindrome_2(ll))

ll.fill([ch for ch in 'i did did i'.replace(' ', '')])
print(ll, '->', is_palindrome_1(ll))
print(ll, '->', is_palindrome_2(ll))

ll.fill([ch for ch in 'dd'.replace(' ', '')])
print(ll, '->', is_palindrome_1(ll))
print(ll, '->', is_palindrome_2(ll))

File: CTCI/Interview Questions/C++/2.6_palindrome.cpp
// Implement a function to check if a linked list is a palindrome.

// Page 64

#include <iostream>
#include <vector>
using namespace std;

struct Node{
    char data;
    Node * next = NULL;
};

class LinkedList{
private:
    Node * head = NULL;
public:
    void append(Node * new_node){
        if(head == NULL)
            head = new_node;
        else{
            Node * current = head;
            while(current -> next)
                current = current -> next;
            current -> next = new_node;
        }
    }
    void append(char data){
        Node * new_node = new Node;
        new_node -> data = data;
        if(head == NULL)
            head = new_node;
        else{
            Node * current = head;
            while(current -> next)
                current = current -> next;
            current -> next = new_node;
        }
    }
    void fill(vector <char> elements){
        if(head) head = NULL; // Doing this just to make the main function cleaner
        for(auto element : elements)
            append(element);
    }
    void fill(string str){
        vector <char> pal_vect (str.begin(), str.end());
        fill(pal_vect);
    }
    void print(){
        Node * current = head;
        while(current){
            cout << current -> data;
            current = current -> next;
        }
    }
    Node * get_head(){
        return head;
    }
    void set_head(Node * new_head){
        head = new_head;
    }
};

// Solution 1
// Using a stack store store the first half of the palindrome, then popping each and checking with latter half of linked list.
// Solution assumes that size of linked list is not known.
// Time Complexity: O(n), Space Complexity: O(n)
bool is_palindrome_1(LinkedList ll){
    Node * head = ll.get_head();
    // If string is less than 1 character long
    if(!head || !head -> next)
        return true;
    // If string is two characters long
    if(head -> data == head -> next -> data && !(head -> next -> next))
        return true;
    Node * slow = head;
    Node * fast = head;
    vector <char> stack;
    while(true){
        stack.push_back(slow -> data);
        slow = slow -> next;
        fast = fast -> next -> next;
        // If character has an odd number of characters
        if(!fast -> next){
            slow = slow -> next;
            break;
        }
        // If character has an even number of characters
        else if(fast -> next && !fast -> next -> next){
            stack.push_back(slow -> data);
            slow = slow -> next;
            break;
        }
    }
    while(slow){
        if(slow -> data != stack.back())
            return false;
        stack.pop_back();
        slow = slow -> next;
    }
    return true;
}

// I could've used a copy constructor but I'm still a noob in C++ so that didn't work out very well
LinkedList clone_ll(LinkedList ll){
    LinkedList rev_ll;
    Node * current = ll.get_head();
    while(current){
        Node * new_node = new Node;
        new_node -> data = current -> data;
        rev_ll.append(new_node);
        current = current -> next;
    }
    return rev_ll;
}

LinkedList reverse_and_clone_ll(LinkedList ll){
    LinkedList rev_ll = clone_ll(ll);
    Node * current = rev_ll.get_head();
    Node * prev = NULL;
    Node * next;
    while(current){
        next = current -> next;
        current -> next = prev;
        prev = current;
        current = next;
    }
    rev_ll.set_head(prev);
    return rev_ll;
}

bool compare_linked_lists(LinkedList ll, LinkedList rev_ll){
    Node * first = ll.get_head();
    Node * second = rev_ll.get_head();
    Node * fast = first;
    while(first && second){
        if(first -> data != second -> data) return false;
        first = first -> next;
        second = second -> next;
        if(fast -> next && fast -> next -> next)
            fast = fast -> next -> next;
        else break;
    }
    return true;
}

// Solution 2
// Reversing linked list and check first half of each list
// Time Complexity: O(n), Space Complexity: O(n)
bool is_palindrome_2(LinkedList ll){
    Node * head = ll.get_head();
    // If string is less than 1 character long
    if(!head || !head -> next)
        return true;
    // If string is two characters long
    if(head -> data == head -> next -> data && !(head -> next -> next))
        return true;
    LinkedList rev_ll = reverse_and_clone_ll(ll);
    return compare_linked_lists(ll, rev_ll);
}

int main(void){
    LinkedList ll;

    ll.fill("red rum sir is murder");
    ll.print();
    cout << " " << is_palindrome_1(ll);
    cout << " " << is_palindrome_2(ll) << endl;

    ll.fill("hello world lol");
    ll.print();
    cout << " " << is_palindrome_1(ll);
    cout << " " << is_palindrome_2(ll) << endl;

    ll.fill("i did did i");
    ll.print();
    cout << " " << is_palindrome_1(ll);
    cout << " " << is_palindrome_2(ll) << endl;

    ll.fill("dd");
    ll.print();
    cout << " " << is_palindrome_1(ll);
    cout << " " << is_palindrome_2(ll) << endl;
}

// Why did this reach 180 lines? Am I doing something wrong?
